vehicle/mkv/software/air_control/air.c

```diff
@@ -1,499 +1,429 @@
-		}
-}
+static void state_machine_run(void) {
+    if (air_control_critical.air_fault != FAULT_NONE) {
+        air_control_critical.air_state = FAULT;
+    }
 
-void sendShutdownSenseCANMessage (void) {
-	msgShutdownSense[MSG_INDEX_SS_TSCONN] = bit_is_set(sFlag, FLAG_SS_TSCONN) ? 0xff : 0x00;
-	msgShutdownSense[MSG_INDEX_SS_HVD] = bit_is_set(sFlag, FLAG_SS_HVD) ? 0xff : 0x00;
-	msgShutdownSense[MSG_INDEX_SS_IMD] = bit_is_set(sFlag, FLAG_SS_IMD) ? 0xff : 0x00;
-	msgShutdownSense[MSG_INDEX_SS_BMS_STATUS] = bit_is_set(sFlag, FLAG_SS_BMS_STATUS) ? 0xff : 0x00;
-	msgShutdownSense[MSG_INDEX_SS_IMD_STATUS] = bit_is_set(sFlag, FLAG_SS_IMD_STATUS) ? 0xff : 0x00;
-	msgShutdownSense[MSG_INDEX_SS_TSMS] = bit_is_set(sFlag, FLAG_SS_TSMS) ? 0xff : 0x00;
-	CAN_transmit(MOB_BROADCAST_SS,
-								CAN_ID_AIR_CONTROL_SENSE,
-								CAN_LEN_AIR_CONTROL_SENSE,
-								msgShutdownSense);
+    switch (air_control_critical.air_state) {
+        case IDLE: {
+            /*
+             * Idle until shutdown circuit is closed
+             */
+            if (air_control_critical.ss_tsms) {
+                air_control_critical.air_state = SHUTDOWN_CIRCUIT_CLOSED;
+            }
+        } break;
+        case SHUTDOWN_CIRCUIT_CLOSED: {
+            static bool once = true;
+
+            if (once) {
+                timer_set = get_time();
+                once = false;
+            }
+
+            if (get_time() - timer_set < 200) {
+                if (air_control_critical.air_n_status) {
+                    air_control_critical.air_state = PRECHARGE;
+                    once = true;
+                    return;
+                } else {
+                    return;
+                }
+            } else {
+                set_fault(FAULT_SHUTDOWN_IMPLAUSIBILITY);
+                once = true;
+                return;
+            }
+        } break;
+        case PRECHARGE: {
+            // Start precharge
+            gpio_set_pin(PRECHARGE_CTL);
+
+            // Get pack voltage to compare with MC voltage
+            int16_t pack_voltage = 0;
+            int16_t motor_controller_voltage = 0;
+            int rc;
+
+            rc = get_bms_voltage(&pack_voltage);
+
+            if (rc != 0) {
+                set_fault(FAULT_CAN_BMS_TIMEOUT);
+                return;
+            }
+
+            // Set correct scale for pack voltage
+            pack_voltage = pack_voltage * 10;
+
+            static bool once = true;
+
+            if (once) {
+                timer_set = get_time();
+                once = false;
+            }
+
+            if (get_time() - timer_set < 2000) {
+                rc = get_motor_controller_voltage(&motor_controller_voltage);
+                if (rc != 0) {
+                    set_fault(FAULT_CAN_MC_TIMEOUT);
+                    once = true;
+                    return;
+                }
+
+                // Set correct scale for MC voltage
+                motor_controller_voltage = motor_controller_voltage / 10;
```

Comment:
Shouldn't this scaling happen in dbc world? Do we define a dbc for signals from motor controller?

Reply:
No, the DBC only scales values when you decode CAN on your laptop. The DBC doesn't do anything with the firmware.

Reply:
Eh, I don't think so, because it is better to encode the offsets and scales in the constants that we compare against.